Introduce Messageable and Transferable for postMessage

nikeee opened the pull request:
This PR addresses https://github.com/Microsoft/TypeScript/issues/25176

Relevant MDN pages:
https://developer.mozilla.org/en-US/docs/Web/API/BroadcastChannel/postMessage
https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage
https://developer.mozilla.org/en-US/docs/Web/API/Client/postMessage
https://developer.mozilla.org/en-US/docs/Web/API/Worker/postMessage
https://developer.mozilla.org/en-US/docs/Web/API/MessagePort/postMessage
https://developer.mozilla.org/en-US/docs/Web/API/DedicatedWorkerGlobalScope/postMessage

Map and Set do not seem to be available in `dom.generated.d.ts`. How would one implement the `Messageable` type?

mhegazy commented:
`Map` and `Set` are defined in the ES6 lib file. you will need to split the declaration into two overloads, and have the `Map`/`Set` ones in `dom.iterable.generated.d.ts`.

mhegazy commented:
@saschanaz can you please review.

nikeee commented:
> Map and Set are defined in the ES6 lib file. you will need to split the declaration into two overloads, and have the Map/Set ones in dom.iterable.generated.d.ts

How do I add types that sould only be included in `dom.iterable.generated.d.ts`.

mhegazy commented:
see https://github.com/Microsoft/TSJS-lib-generator/blob/master/src/index.ts#L24

nikeee commented:
I couldn't find any way to include different type defs without changing the emitter code. I removed the `Messageable` type def for now.

I added OffscreenCanvas to `Transferable` since it's also Transferable [according to the spec IDL](https://html.spec.whatwg.org/multipage/canvas.html#the-offscreencanvas-interface)

saschanaz commented on inputfiles/overridingTypes.json:
```diff
@@ -2318,6 +2337,58 @@
+                        "postMessage": {
+                            "name": "postMessage",
+                            "override-signatures": [
+                                "postMessage(message: any, transferList?: Transferable[]): void"
```
Why `transferList` instead of `transfer` here?

nikeee replied:
I named it `transferList` because it is landed so in the MDN docs.

https://developer.mozilla.org/en-US/docs/Web/API/Worker/postMessage

I can change it to `transfer` if it's preferred.

nikeee replied:
I changed it to `transfer`.
